### ISSUE TITLE
Allow to set individual calendars as read-only

### ISSUE THREAD

**Issue by ralf1070** [jobisoft/TbSync#85]
## Your environment

TbSync version: 0.7.7.20
Thunderbird version: 52.8.0 (64-Bit)

[x] Yes, I have installed the latest available (beta) version from 
https://github.com/jobisoft/TbSync/releases 
and my issue is not yet fixed, I can still reproduce it.

## Expected behavior
Flawless sync.

## Actual behavior
I'm subscribed to several calendars and address books on a kopano server. This works quite good but after some days TbSync starts to report errors when synchronizing calendars:

Sync failed. Server responded with status <Sync.Collections.Collection.Responses.Change[0].S = 9>.

Restarts of Thunderbird does not affect this problem. 
Unsubscribing/Subscribing such a calendar fixes it.

Note: it may be related or not - most of the time, most of these subscribed calendars are deactivated in the thunderbird calendar view (as these are calendars of co-workers, which are only relevant for me when I have to look up dates with them). For whatever reason at least some of their appointments show up in my reminder popup even when these calendars are deactivated. The log seems to suggest that TbSync tries to write local changes of the affected calendars to our server. This is strange as I can guarantee, that I did not do any changes to these calendars at all - maybe except dismissing irrelevant reminders.

**Comment by jobisoft** [comment]
can you use the new "revert local changes" option to prevent modifications on your co-worker calendars.

Does that fix your problem?

**Comment by jobisoft** [comment]
Status (9) means "User account could be out of disk space." So yeah, it looks like the server is not able to write changes to your co-workers calendar. The "revert local changes" option should do the trick for you.

You need to disable your account to be able to edit the settings.

**Comment by jobisoft** [comment]
Oh, I implemented that as a global option, not per calendar/folder. You can still test it, but you will write-protect the entire account... I have to think about that for a while...

**Comment by ralf1070** [comment]
Wow - fast reaction.

I first have to wait until the error comes up again. Then I will try this option to see if it helps.

What is fascinating for me is the fact that TbSync wants to write at all, because I did not do any active modifications to my co-workers calendars. The other fascinating fact is that I am able to create or modify appointments in my co-workers calendars without problems - except in the case stated above. Just an idea: is it possible that TbSync tries to update entries which are modified by another client and there is a conflict which is not correctly resolved?

**Comment by jobisoft** [comment]
thunderbird somehow modifies alarms/reminder states which I cannot stop. This will not actually change data, but the item is marked as "modified" and TbSync wants to sync. Have not yet found away to check that condition (and undo the mod mark).

If you wait till the error shows up again, that option will not help anymore. If you want to enable that option, you have to disable the account, which deletes the local calendar and after re-enable, you get the current server version. Your local mods are gone. 

Thinking about it, the test will not gain us any insights, because the "send code" (which triggered the error) is never executed, so the error cannot show up.

I will think about handling error 9 differently...

**Comment by ralf1070** [comment]
> I will think about handling error 9 differently...

Would be great.

BTW: I am curious: what means "User account could be out of disk space."? This is a strange king of response - is the account out of space or not? Do you have an idea what a client, which receives such a result, is supposed to do?

**Comment by jobisoft** [comment]
`
Sync failed. Server responded with status <Sync.Collections.Collection.Responses.Change[0].S = 9>.
`
Your server reports status 9 and in this link
https://msdn.microsoft.com/en-us/library/gg675457(v=exchg.80).aspx

you can see what that means. Since there is no dedicated "user has no write permissions" status, Kopano is sending this status, to indicate that it could not write the changes. I guess.

**Comment by ralf1070** [comment]
I see. Thanks a lot.

**Comment by jobisoft** [comment]
I just release beta .22 which ignores status 9.

This is just a quick fix for you. The server will still send status 9 on local changes to these calendars, where you have no write permissions, but TbSync will just "smile and wave" and not report any error.

The global "revert local changes" option seems not to work with Microsoft servers, so do not use it.

In the long run, I will implement a read-only option per calendar/addressbook, which is what you actually need. But that will take time. It is not at the top of my todo list. I keep this issue open, as a reminder.

**Comment by ralf1070** [comment]
Great. Thank you very much.

**Comment by zephyrus84** [comment]
Hello, maybe I'm facing the same issue but to be honest it's not correlated to the latest version.

I'm using 0.7.7.21 version and I'm sync calendar, contacts and acrivities.

If I set Sync interval to 0 it gives me always this status when I create an event through Thunderbird:
![schermata del 2018-06-08 09-17-33](https://user-images.githubusercontent.com/40055349/41145632-11558572-6b01-11e8-9946-71e2fe876a89.png)

The message remains until I push "sync" button.

If I set at least 1 as sync interval It get synchronized:
![schermata del 2018-06-08 09-20-16](https://user-images.githubusercontent.com/40055349/41145675-36b9ecea-6b01-11e8-9c5c-120faac26cba.png)

The big problem is that when I create a new event (even with 1 as sync interval) no notification is sent to the partecipant, neither if I manualy push the sync button.

**Comment by jobisoft** [comment]
You describe two different issues:

If the sync interval is set to 0, TbSync will not sync automatically. The message you get is simply "you changed something, you must sync"

If the sync interval is set to 1, TbSync will sync every minute, if you set it to 5, it will sync every 5 minutes...

Now the second issue: Invitations simply do not work yet. Sorry.
https://github.com/jobisoft/TbSync/wiki/What-is-TbSync%3F

**Comment by zephyrus84** [comment]
Thanks a lot for your feeback. I thought it was a correlated problem. Anyway I hope this feature comes soon ;)

**Comment by ralf1070** [comment]
After resubscribing my faulty calendars I still don't have a problem with them. A co-worker, using TB on Linux, just upgraded to TbSync 0.7.7.22 to fix his faulty calendars too but got the following error:

An error occurred when writing to the calendar Gubor! Error code:
MODIFICATION_FAILED. Description:   calCalendarManager.js:966
	calMgrCalendarObserver.prototype.announceError
file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/calendar-js/calCalendarManager.js:966:13
	calMgrCalendarObserver.prototype.onError
file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/calendar-js/calCalendarManager.js:815:9
	notifyFunc
file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/calendar-js/calUtils.js:1240:29
	forEach self-hosted:251:13
	calListenerBag.prototype.notify
file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/calendar-js/calUtils.js:1246:9
	cal.ObserverBag.prototype.notify
resource://calendar/modules/calProviderUtils.jsm:461:16
	cal.ProviderBase.prototype.notifyError
resource://calendar/modules/calProviderUtils.jsm:672:9
	cal.ProviderBase.prototype.notifyOperationComplete
resource://calendar/modules/calProviderUtils.jsm:653:13
	calStorageCalendar.prototype.doModifyItem
file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/components/calStorageCalendar.js:486:13
	calStorageCalendar.prototype.modifyItem/offlineJournalFlagListener.onOperationComplete file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/components/calStorageCalendar.js:477:17
	calStorageCalendar.prototype.getItemOfflineFlag/listener.handleCompletion file:///.../extensions/%7Be2fda1a4-762b-4020-b5ad-a41df1933103%7D/components/calStorageCalendar.js:959:21

**Comment by jobisoft** [comment]
that does not look like a TbSync error message. Can you co-worker disable and re-enable his account? To start "fresh"?

**Comment by jobisoft** [comment]
Ah wait, did your co-worker tried to activate "read-only" support at the calendar itself?

![image](https://user-images.githubusercontent.com/5830621/41152842-2db51642-6b15-11e8-9990-79ea0ebc1a9b.png)

That is not working, because TbSync needs to write to that calendar!

**Comment by ralf1070** [comment]
Yes - he did.

**Comment by ralf1070** [comment]
He disabled readonly and now it works. Thank you.

Without knowing anything about the thunderbird API: is it possible that you use a kind of "UI to calendar API" instead of a "calendar provider API"? For what I'm able to imagine, the read only check box should only suppress **user** writes to calendars. It must not suppress changes of the local calendar model which were triggered externally - so there should be a way to update a calendar model without running into this check. This could also explain why you get modification notifications on calendars which do not have user modifications at all - you get notified about the updates TbSync did when it inserted remote modifications. Just an idea - I don't know your code, I don't know TB-API's.

**Comment by ralf1070** [comment]
Ah I see - was just an idea.

**Comment by tcigler** [comment]
Hi, I also have similar problem. Calendars are NOT read-only and even when I select "revert local changes", I get error. After synchronization is everything OK, but thunderbird shows all (even outdated) events
![selection_115](https://user-images.githubusercontent.com/13203030/46857046-4eafbb80-ce08-11e8-918c-8172b3b17133.png)
When I click Dismiss All, I get the synchonization error with local changes
![selection_116](https://user-images.githubusercontent.com/13203030/46857064-5ec79b00-ce08-11e8-8e7b-faffc38e433d.png)

**Comment by jobisoft** [comment]
yeah that is a loop, the "error" you get is just to inform you, that you have local changes (dismissed alarms) and of course if you sync that via revert, you will be prompted again for dismiss...

however, that should be fixed in TB60, and you should no longer have to dismiss old alarms... what is your environment?

**Comment by tcigler** [comment]
Thanks a lot for the reply. Yes, it's exactly like you wrote. 
I have latest 0.7.16.6 TbSync version after today update (before I had 0.7.16 with the same behavior) on 60.0 Thunderbird (Debian 9 Stable). Interestingly the testing repository does have version 52.9.1 so I shouldn't use the testing then.

**Comment by cincodenada** [comment]
I have been experiencing this loop for a while. I just upgraded to TB60.2.1, with the Ubuntu-provided Calendar package, and am on TbSync 0.7.16, and it still occurs: I set up an account with one-way sync, it syncs and pops up a bunch of old reminders, I dismiss them all, and then it goes into "Account needs to be syncronized to revert local changes (one-way sync)" state.

I've also tried just closing the dialog box, but with no success. Let me know if there are logs I can gather, I can reproduce this pretty reliably still.

**Comment by jobisoft** [comment]
I hear you, this is a bug I need to fix, I will. Sorry for your hassle.

**Comment by cincodenada** [comment]
Thank you! I didn't intend my comment to be impatient, I know there's a lot to do and I appreciate all the work you're putting in here, you've already made my Thunderbird usage much improved. I just meant to offer another report with my specific details. If there's something I can to do help, do let me know!

**Comment by jobisoft** [comment]
:-) all fine. I do not need another report because it is a well known but not easy to fix bug. I have no good idea *how* to fix it yet and currently my mind is on other things. It is the 3rd point on my list.

**Comment by jobisoft** [comment]
The latest beta has options to set individual folders (calendars, adress books) as read-only.

![image](https://user-images.githubusercontent.com/5830621/51067204-3a9a7680-1610-11e9-803f-e7ba8ffaf68c.png)

EAS does not provide a way to detect read-only access, so you have to set it by yourself.

To get the latest beta, install all(!) XPI modules found here:
https://tbsync.jobisoft.de/beta

Feedback would be nice.

**Comment by jobisoft** [comment]
This is pending for release. No one wants to try?

**Comment by cincodenada** [comment]
Hey, I gave this a run today as I was back in the office. I installed the beta XPIs, and so far it's working a dream, this is a great improvement, the warning message is nice. With the beta version, I can now enable the periodic synchronization without be prompted to re-dismiss everything every sync, and the state settles at Idle instead of "Changes need reverting."

So, this gets a big thumbs up from me! I still get prompted occasionally to dimsiss things in the near future, I haven't kept track of whether it happens more than once per event, and it doesn't happen every sync, so that's just fine.
